README.md
cloud-canary
============

Bunch of Python scripts that perform various checks to test cloud provider availability.

- cloud-canary

The goal of the script is to schedule a recurring test on a given cloud provider (in our case exoscale) by creating a "canary instance", try to connect to it using SSH and finally destroy it. If any exception occur during the process, an alarm is sent to Riemann monitoring system. Exec time metric is also sent to Riemann to get the latency.

- api-canary

The goal of the script is to schedule a recurring test on a given cloud provider (in our case exoscale) by performing a test api call (list size). If any exception occur during the process, an alarm is sent to Riemann monitoring system. Exec time metric is also sent to Riemann to get the latency.

- s3-canary

The goal of the script is to schedule a recurring test on a given s3 provider by creating a file, read it's content and delete it. If any exception occur during the process, an alarm is sent to Riemann monitoring system. Exec time metric is also sent to Riemann to get the latency.

These scripts relies on Apache Libcloud (https://libcloud.apache.org/) and Riemann (http://riemann.io/) and boto.

Requirements
------------

You may install the requirements using the following commands:

```
pip install apache-libcloud
pip install paramiko (may require python-dev package)
pip install bernhard
pip install boto
```

Scripts tested only on Ubuntu.


debian/changelog
exoscale-cloud-canary (1.0-1) unstable; urgency=medium

  * Initial release.

 -- Exoscale <ops@exoscale.ch>  Fri, 23 Mar 2018 10:00:00 +0000


api-canary.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Loic Lambiel ©
# License MIT

import sys
import argparse
import logging
import logging.handlers
import time
import socket

try:
    from libcloud.compute.types import Provider
    from libcloud.compute.providers import get_driver
except ImportError:
    print ("It look like libcloud module isn't installed. Please install it using pip install apache-libcloud")
    sys.exit(1)


try:
    import bernhard
except ImportError:
    print ("It look like riemann client (bernard) isn't installed. Please install it using pip install bernhard")
    sys.exit(1)

try:
    from configparser import ConfigParser
except ImportError:  # python 2
    from ConfigParser import ConfigParser

logfile = "/var/log/api-canary.log"
logging.basicConfig(format='%(asctime)s %(pathname)s %(levelname)s:%(message)s', level=logging.DEBUG, filename=logfile)
logging.getLogger().addHandler(logging.StreamHandler())


def main():
    parser = argparse.ArgumentParser(description='This script perform a list_size API query on exoscale public cloud. If any error occur during the process, an alarm is being sent to riemann monitoring. time metric is also sent to riemann')
    parser.add_argument('-version', action='version', version='%(prog)s 1.0, Loic Lambiel, exoscale')
    parser.add_argument('-acskey', help='Cloudstack API user key', required=True, type=str, dest='acskey')
    parser.add_argument('-acssecret', help='Cloudstack API user secret', required=True, type=str, dest='acssecret')
    args = vars(parser.parse_args())
    return args


def list_size(args):
    API_KEY = args['acskey']
    API_SECRET_KEY = args['acssecret']

    cls = get_driver(Provider.EXOSCALE)
    driver = cls(API_KEY, API_SECRET_KEY)

    logging.info('Performing query')

    size = driver.list_sizes()

    micro = False

    for item in size:
        if item.name == 'Micro':
            micro = True

    if micro is False:
        raise Exception("API call did not returned Micro instance type. This means the API isn't working correctly")


# main
if __name__ == "__main__":
    args = main()

    conf = ConfigParser()
    conf.read(("/etc/bernhard.conf",))
    client = bernhard.SSLClient(host=conf.get('default', 'riemann_server'),
                                port=int(conf.get('default', 'riemann_port')),
                                keyfile=conf.get('default', 'tls_cert_key'),
                                certfile=conf.get('default', 'tls_cert'),
                                ca_certs=conf.get('default', 'tls_ca_cert'))

    start_time = time.time()
    try:
        list_size(args)
        exectime = time.time() - start_time
        host = socket.gethostname()
        client.send({'host': host,
                     'service': "api_canary.exectime",
                     'state': 'ok',
                     'tags': ['duration'],
                     'ttl': 600,
                     'metric': exectime})
        client.send({'host': host,
                     'service': "api_canary.check",
                     'state': 'ok',
                     'tags': ['api_canary.py', 'duration'],
                     'ttl': 600,
                     'metric': 0})

        logging.info('Script completed successfully')

    except Exception as e:
        logging.exception("An exception occured. Exception is: %s", e)
        host = socket.gethostname()
        exectime = 61
        txt = 'An exception occurred on api_canary.py: %s. See logfile %s for more info' % (e, logfile)
        client.send({'host': host,
                     'service': "api_canary.check",
                     'description': txt,
                     'state': 'critical',
                     'tags': ['api_canary.py', 'duration'],
                     'ttl': 600,
                     'metric': 1})
        client.send({'host': host,
                     'service': "api_canary.exectime",
                     'state': 'ok',
                     'tags': ['duration'],
                     'ttl': 600,
                     'metric': exectime})
        raise


cloud-canary.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Loic Lambiel ©
# License MIT

import argparse
import logging
import logging.handlers
import pprint
import socket
import sys
import time


try:
    from libcloud.compute.types import Provider
    from libcloud.compute.providers import get_driver
    from libcloud.compute.deployment import ScriptDeployment
    from libcloud.compute.deployment import MultiStepDeployment
    from libcloud.compute.base import NodeImage
except ImportError:
    print("It look like libcloud module isn't installed. Please install it "
          "using pip install apache-libcloud")
    sys.exit(1)


try:
    import bernhard
except ImportError:
    print("It look like riemann client (bernard) isn't installed. Please "
          "install it using pip install bernhard")
    sys.exit(1)

try:
    from configparser import ConfigParser
except ImportError:  # python 2
    from ConfigParser import ConfigParser


def main():
    parser = argparse.ArgumentParser(description='''
This script spawn an instance on exoscale public cloud and execute a dummy
command thru SSH. If any error occur during the process, an alarm is being
sent to riemann monitoring''')
    parser.add_argument('-version', action='version',
                        version='%(prog)s 1.0, Loic Lambiel, exoscale')
    parser.add_argument('-acskey', help='Cloudstack API user key',
                        required=True, type=str, dest='acskey')
    parser.add_argument('-acssecret', help='Cloudstack API user secret',
                        required=True, type=str, dest='acssecret')
    parser.add_argument('-zonename', help='Cloudstack zone name',
                        required=True, type=str, dest='zonename')
    parser.add_argument('-alertstate',
                        help='Alert state to raise if the test fails',
                        required=False, type=str, default='critical',
                        dest='state')
    parser.add_argument('-endpoint', help='The API endpoint', required=False,
                        type=str, default='api.exoscale.ch', dest='endpoint')
    parser.add_argument('-template',
                        help='Template name to use for the deployment',
                        required=False, type=str,
                        default='Linux Ubuntu 16.04 LTS 64-bit 10G',
                        dest='template')
    parser.add_argument('-offering',
                        help='Service offering to use for the deployment',
                        required=False, type=str,
                        default='Micro',
                        dest='offering')
    args = vars(parser.parse_args())
    return args


def deploy_instance(args):
    api_key = args['acskey']
    secret_key = args['acssecret']
    zonename = args['zonename']
    endpoint = args['endpoint']
    template = args['template']
    offering = args['offering']

    logging.basicConfig(
        format='%(asctime)s %(pathname)s %(levelname)s:%(message)s',
        level=logging.DEBUG,
        filename=logfile)
    logging.getLogger().addHandler(logging.StreamHandler())

    cls = get_driver(Provider.EXOSCALE)
    driver = cls(api_key, secret_key, host=endpoint)

    location = [location for location in driver.list_locations()
                if location.name.lower() == zonename.lower()][0]

    size = [size for size in driver.list_sizes() if size.name == offering][0]
    images = [i for i in driver.list_images()
              if template.lower() in i.extra['displaytext'].lower()]
    images = sorted(images, key=lambda i: i.extra['displaytext'], reverse=True)
    image = NodeImage(id=images[0].id, name=images[0].name, driver=driver)

    name = 'canary-check-' + location.name.lower()

    if endpoint != 'api.exoscale.ch':
        name += '-pp'

    # check if a previous canary exists
    nodes = driver.list_nodes()
    for n in nodes:
        if name in n.name:
            raise Exception('Instance with same name already exists !')

    script = ScriptDeployment('echo Iam alive !')
    msd = MultiStepDeployment([script])

    logging.info('Deploying instance %s', name)

    node = driver.deploy_node(name=name, image=image, size=size, timeout=300,
                              location=location, ssh_username='ubuntu',
                              deploy=msd)
    logging.debug(pprint.pformat(node))

    nodename = str(node.name)
    nodeid = str(node.uuid)
    nodeip = str(node.public_ips)
    logging.info('Instance successfully deployed : %s, %s, %s', nodename,
                 nodeid, nodeip)
    # The stdout of the deployment can be checked on the `script` object
    logging.debug(pprint.pformat(script.stdout))

    logging.info('Successfully executed echo command thru SSH')
    logging.info('Destroying the instance now')
    # destroy our canary node
    driver.destroy_node(node)

    logging.info('Successfully destroyed the instance %s', name)
    logging.info('Script completed')


# main
if __name__ == "__main__":
    args = main()
    zonename = args['zonename']
    state = args['state']
    endpoint = args['endpoint']
    if endpoint == "ppapi.exoscale.ch":
        env = "pp"
    else:
        env = "prod"
    logfile = "/var/log/cloud-canary-{}-{}.log".format(env, zonename)
    conf = ConfigParser()
    conf.read(("/etc/bernhard.conf",))
    if endpoint != 'api.exoscale.ch':
        zonename += '-pp'

    client = bernhard.SSLClient(host=conf.get('default', 'riemann_server'),
                                port=int(conf.get('default', 'riemann_port')),
                                keyfile=conf.get('default', 'tls_cert_key'),
                                certfile=conf.get('default', 'tls_cert'),
                                ca_certs=conf.get('default', 'tls_ca_cert'))
    start_time = time.time()
    try:
        deploy_instance(args)
        exectime = time.time() - start_time
        host = socket.gethostname()
        client.send({'host': host,
                     'service': "Cloud_canary-" + zonename + ".exectime",
                     'state': 'ok',
                     'tags': ['duration'],
                     'ttl': 3800,
                     'metric': exectime})
        client.send({'host': host,
                     'service': "Cloud_canary-" + zonename + ".check",
                     'state': 'ok',
                     'tags': ['cloud_canary.py', 'duration'],
                     'ttl': 3800,
                     'metric': 0})
    except Exception as e:
        logging.exception("An exception occured. Exception is: %s", e)
        host = socket.gethostname()
        txt = ("An exception occurred on cloud_canary.py: {}. See logfile {} "
               "for more info").format(e, logfile)
        client.send({'host': host,
                     'service': "Cloud_canary-" + zonename + ".check",
                     'description': txt,
                     'state': state,
                     'tags': ['cloud_canary.py', 'duration'],
                     'ttl': 3800,
                     'metric': 1})
        raise


download-canary.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Loic Lambiel ©
# License MIT

import argparse
import logging
import logging.handlers
import socket
import sys
import time
import urllib2


try:
    import bernhard
except ImportError:
    print("It look like riemann client (bernard) isn't installed. "
          "Please install it using pip install bernhard")
    sys.exit(1)


try:
    from configparser import ConfigParser
except ImportError:  # python 2
    from ConfigParser import ConfigParser


def main():
    parser = argparse.ArgumentParser(description='''
This script create download a small file from a given URL.
If any error occur during the process, an alarm is being sent
to riemann monitoring. time metric is also sent to riemann''')
    parser.add_argument('-version',
                        action='version',
                        version='%(prog)s 1.0, Loic Lambiel, exoscale')
    parser.add_argument('-url',
                        help='URL of the file to be downloaded',
                        required=True,
                        type=str,
                        dest='url')
    parser.add_argument('-alertstate',
                        help='Alert level to raise if the test fails',
                        required=False,
                        type=str,
                        default='critical',
                        dest='state')
    parser.add_argument('-env',
                        help='Environnement used in the riemann service',
                        required=True,
                        type=str,
                        dest='env')
    args = vars(parser.parse_args())
    return args


def downloadtest(args):
    URL = args['url']
    ENV = args['env']

    timeout = 30
    socket.setdefaulttimeout(timeout)

    logging.info('Downloading file for env %s', ENV)

    req = urllib2.Request(URL)
    response = urllib2.urlopen(req)
    response.read()

    logging.info('Download completed for env %s', ENV)


# main
if __name__ == "__main__":
    args = main()
    ENV = args['env']
    URL = args['url']
    sos_domain = URL.split('/')[2]
    conf = ConfigParser()
    conf.read(("/etc/bernhard.conf",))
    logfile = "/var/log/download-canary-{}.log".format(sos_domain)
    logging.basicConfig(format=('%(asctime)s %(pathname)s '
                                '%(levelname)s:%(message)s'),
                        level=logging.DEBUG,
                        filename=logfile)
    logging.getLogger().addHandler(logging.StreamHandler())
    logging.getLogger('boto').setLevel(logging.CRITICAL)

    client = bernhard.SSLClient(host=conf.get('default', 'riemann_server'),
                                port=int(conf.get('default', 'riemann_port')),
                                keyfile=conf.get('default', 'tls_cert_key'),
                                certfile=conf.get('default', 'tls_cert'),
                                ca_certs=conf.get('default', 'tls_ca_cert'))

    logging.info('Test started for env %s', ENV)
    ENV = args['env']
    state = args['state']
    checkservice = "{}.download_canary.check".format(sos_domain)
    exectimeservice = "{}.download_canary.exectime".format(sos_domain)
    start_time = time.time()
    host = socket.gethostname()
    try:
        downloadtest(args)
        exectime = time.time() - start_time
        client.send({'host': host,
                     'service': exectimeservice,
                     'state': 'ok',
                     'tags': ['duration', ENV],
                     'ttl': 600,
                     'metric': exectime})
        client.send({'host': host,
                     'service': checkservice,
                     'state': 'ok',
                     'tags': ['download_canary.py', 'duration', ENV],
                     'ttl': 600,
                     'metric': 0})

        logging.info('Script completed successfully for env %s', ENV)

    except Exception as e:
        logging.exception("An exception occured. Exception is: %s", e)
        exectime = 61
        txt = "An exception occurred on download_canary.py: {}. \
               See logfile {} for more info".format(e, logfile)
        client.send({'host': host,
                     'service': checkservice,
                     'description': txt,
                     'state': state,
                     'tags': ['download_canary.py', 'duration', ENV],
                     'ttl': 600,
                     'metric': 1})
        client.send({'host': host,
                     'service': exectimeservice,
                     'state': 'ok',
                     'tags': ['duration', ENV],
                     'ttl': 600,
                     'metric': exectime})
        raise


s3-canary.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Loic Lambiel ©
# License MIT

import argparse
import logging
import logging.handlers
import socket
import sys
import time

try:
    import boto
    import boto.s3.connection
    from boto.s3.key import Key
except ImportError:
    print("It look like boto module isn't installed. \
          Please install it using pip install boto")
    sys.exit(1)


try:
    import bernhard
except ImportError:
    print("It look like riemann client (bernard) isn't installed. \
          Please install it using pip install bernhard")
    sys.exit(1)


try:
    from configparser import ConfigParser
except ImportError:  # python 2
    from ConfigParser import ConfigParser


class CustomError(Exception):

    def __init__(self, arg):
        self.msg = arg


def main():
    parser = argparse.ArgumentParser(description='''
This script create a file, read it and delete it on a s3 compliant
storage. If any error occur during the process, an alarm is being
sent to riemann monitoring. time metric is also sent to riemann''')
    parser.add_argument('-version',
                        action='version',
                        version='%(prog)s 1.0, Loic Lambiel, exoscale')
    parser.add_argument('-key',
                        help='S3 user key',
                        required=True,
                        type=str,
                        dest='key')
    parser.add_argument('-secret',
                        help='S3 user secret',
                        required=True,
                        type=str,
                        dest='secret')
    parser.add_argument('-host',
                        help='S3 host',
                        required=True,
                        type=str,
                        dest='host')
    parser.add_argument('-bucket',
                        help='S3 bucket',
                        required=True,
                        type=str,
                        dest='bucket')
    parser.add_argument('-env',
                        help='Environnement used by the riemann service',
                        required=True,
                        type=str,
                        dest='env')
    parser.add_argument('-alertstate',
                        help='Alert level to raise if the test fails',
                        required=False,
                        type=str,
                        default='critical',
                        dest='state')
    args = vars(parser.parse_args())
    return args


def s3test(args):
    KEY = args['key']
    SECRET = args['secret']
    HOST = args['host']
    BUCKET = args['bucket']
    ENV = args['env']

    logging.basicConfig(format=('%(asctime)s %(pathname)s '
                                '%(levelname)s:%(message)s'),
                        level=logging.DEBUG,
                        filename=logfile)
    logging.getLogger().addHandler(logging.StreamHandler())
    logging.getLogger('boto').setLevel(logging.DEBUG)
    logging.info('Test started for env %s', ENV)

    conn = boto.connect_s3(
        aws_access_key_id=KEY,
        aws_secret_access_key=SECRET,
        host=HOST,
        calling_format=boto.s3.connection.OrdinaryCallingFormat(),
    )

    logging.info('Creating bucket %s', BUCKET)
    bucket = conn.create_bucket(BUCKET)
    logging.info('Bucket create for env %s', ENV)

    k = Key(bucket)
    k.key = 's3-canary'
    logging.info('Writing file')
    k.set_contents_from_string('This is a test of S3')
    logging.info('Write done for env %s', ENV)
    time.sleep(3)
    logging.info('Reading back the file for env %s', ENV)

    if k.get_contents_as_string() != "This is a test of S3":
        raise CustomError('Failed to read back the created file !')

    logging.info('File has been read back for env %s', ENV)
    logging.info('Deleting bucket for env %s', ENV)
    bucket.delete_key(k)
    logging.info('Bucket deleted for env %s', ENV)


# main
if __name__ == "__main__":
    args = main()
    region = args['host'].split('.')[0]
    logfile = "/var/log/s3-canary-{}.log".format(region)
    conf = ConfigParser()
    conf.read(("/etc/bernhard.conf",))

    client = bernhard.SSLClient(host=conf.get('default', 'riemann_server'),
                                port=int(conf.get('default', 'riemann_port')),
                                keyfile=conf.get('default', 'tls_cert_key'),
                                certfile=conf.get('default', 'tls_cert'),
                                ca_certs=conf.get('default', 'tls_ca_cert'))
    ENV = args['env']
    HOST = args['host']
    state = args['state']
    exectimeservice = "{}.s3_canary.exectime".format(HOST)
    checkservice = "{}.s3_canary.check".format(HOST)
    start_time = time.time()
    try:
        s3test(args)
        exectime = time.time() - start_time
        host = socket.gethostname()
        client.send({'host': host,
                     'service': exectimeservice,
                     'state': 'ok',
                     'tags': ['duration', ENV],
                     'ttl': 600,
                     'metric': exectime})
        client.send({'host': host,
                     'service': checkservice,
                     'state': 'ok',
                     'tags': ['s3_canary.py', 'duration', ENV],
                     'ttl': 600,
                     'metric': 0})

        logging.info('Script completed successfully for env %s', ENV)

    except Exception as e:
        logging.exception("An exception occured. Exception is: %s", e)
        host = socket.gethostname()
        exectime = 61
        txt = "An exception occurred on s3_canary.py: {}. \
               See logfile {} for more info".format(e, logfile)
        client.send({'host': host,
                     'service': checkservice,
                     'description': txt,
                     'state': state,
                     'tags': ['s3_canary.py', 'duration', ENV],
                     'ttl': 600,
                     'metric': 1})
        client.send({'host': host,
                     'service': exectimeservice,
                     'state': 'ok',
                     'tags': ['duration', ENV],
                     'ttl': 600,
                     'metric': exectime})
        raise
